Feature/environment revision

DominicDirkx opened the pull request:
As requested during the Tudat meeting of Friday June 19, I'm opening a pull request of my ongoing efforts to improve the documentation on docs.tudat.space.

Pages I'm particularly keen to get input on are:

- Setting up custom models: https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/environment_setup/architecture/custom_settings.rst
- Interacting with the environment during propagation: https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/environment_setup/architecture/environment_during_propagation.rst
- Specific environment model considerations (ideas for a new name? someone had a good one during the meeting... ), second part of: https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst
- Overall flow of state propagation: https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst
- Overview of setup of submodules (not mentioned during meeting) https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_getting_started/submodules.rst
- Guide on how to modify old aerodynamic/thrust guidance code to the new setup https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/environment_setup/thrust_refactor/thrust_refactor.rst

However, I am interested in *any* input you may have!

MiguelAvillez commented:
**Interacting with the environment during propagation:**
- Some typos: wriing, necessarilly, convert **forced** to accelerations, settigs, teh, atmospher, exracted
- Regarding the inertial vs gravitational mass I would add some additional comment/reference explaining the difference between the two... mostly because I don't actually know what the difference is. 
- I would switch the title to "Interacting the environment during propagation" (is it too long?), as to me just "environment during propagation" isn't super instructive
- In general about the page, is the idea for it to be just a few examples of stuff that can be retrieved from the `Body` class?

**Specific environment model considerations:**
- Typos: calcualting, neede, gravitional
- "For a non-spherical central body, the current orientation of **a body** has an indirect influence on the altitude at which a vehicle with a given inertial state is located" - clarify that "a body" is the central body

**Overview of setup of submodules:**
- Typos: functitons, These objects may have various interdependencies **what** are difficult to manually implement, teh 
- "The usability of separate functions/classes in this submodule outside of the Tudatpy framework is very limited." Is this necessary? Took me a second to realize what it refers to, and I don't think it adds much to the paragraph

Regarding these 3 pages, the level of detail seems appropriate and I think the explanations are quite clear :)
I'll add some comments (if need be) about the other 2 pages later

DominicDirkx commented:
Hi Miguel, thanks for the thorough read-through! I've processed each of your comments, with one exception "The usability of separate functions/classes in this submodule outside of the Tudatpy framework is very limited." This line is here to indicate that the functionality here is not made for (for instance) using only an ephemeris, only an integrator, only a gravity field, etc. outside of Tudat. The functionality generally comes as a whole (unlike things in the math or astro submodule). I've added a line to this to maybe make it more clear. What do you think?

I've also written a new page:
https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/environment_setup/thrust_refactor/thrust_refactor.rst
on how to translate old code to new code (initial post also updated). Your thoughts on that page would also be most welcome :)

MiguelAvillez commented:
> I've processed each of your comments, with one exception "The usability of separate functions/classes in this submodule outside of the Tudatpy framework is very limited." This line is here to indicate that the functionality here is not made for (for instance) using only an ephemeris, only an integrator, only a gravity field, etc. outside of Tudat. The functionality generally comes as a whole (unlike things in the math or astro submodule). I've added a line to this to maybe make it more clear. What do you think?

Yep, I think it's good now. Though you inserted another small typo: functionaluty

DominicDirkx commented:
@transferorbit Thanks for this rigorous feedback, it's much appreciated :) Apologies for the poor spelling and sentence structure, if you see this in other files, feel free to modify and push the changes (or continue doing as you did here, whatever works) Below are a few thoughts, for us to discuss if necessary
* In principle, the user guide assumed familiarity with Python. **Important question:** is the terminology I am using (class, object, member, member function) also correct in Python?
* _Check all the 'body' words. Do you want body, Body, or `Body`?_ That's a good question, which I was struggling with when writing. On the one hand, I would like them all to be the same, on the other hand, I think it's good that in certain cases they are links to the Body class in the API reference. **Question**: should they _all_ be links to the API docs?
* _Separately, there is some niggling dissonance for me about the idea of 'models' that are updated during computation. For my feeling (which may need updating, mea culpa), models (in general) are a fixed thing that a simulation is based on. Change the model, and it feels like one is by (my) definition running a different simulation._ This is **exactly** the kind of comments I'm looking for! My way of thinking about these things is so ingrained that I probably have huge blind spots for confusions such as this one. Thanks!

gaffarelj commented on docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/create_body_settings.rst:
```diff
@@ -35,39 +35,56 @@ solar-system bodies. The 'default' settings can be retrieved as follows, using t
-================================
+==============================
 
+Although the default body settings are often very useful, there are various cases where a user will want to override these default settings, or where these default settings cannot be use. The manner in which to overrride the default settings can be divided into three categories:
```
The manner in which to ~~overrride~~ *override* the default settings can be divided into three categories:

gaffarelj commented on docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/create_body_settings.rst:
```diff
@@ -89,12 +106,120 @@ Default settings may be overridden as follows:
+          .. literalinclude:: /_src_snippets/simulation/environment_setup/req_create_bodies.cpp
+             :language: cpp
+
+Where the value of the gravitational parameter in the Sun's gravity field is changed to 1.32712440042 :math:`\cdot` 10 :superscript:`20` m :superscript:`3` / s :superscript:`2`. Functionally, this example is identical to the previous one, but it permits different kinds of modifications to be made. It allows only a *single* property of the environment model to be modified, while in the previous ex ...
```
The present example ~~therefor~~ *therefore* allows for more...
...but limits the user to ~~a~~ modifying the properties of the settings...

gaffarelj commented on docs/source/_src_user_guide/state_propagation/environment_setup/thrust_refactor/thrust_refactor.rst:
```diff
@@ -0,0 +1,168 @@
+
+**propagation_setup.thrust.thrust_from_existing_body_orientation** This option is now moot, *by definition* all thrust models now use the existing body orientation. If you were using this option, now additional action is needed.
+
+**propagation_setup.thrust.custom_thrust_orientation**  This option is now replaced with the equivalent :func:`~tudatpy.numerical_simulation.environment_setup.rotation_model.custom_rotation_model` option, with the addition of the frame identiers specifief above. It defines the body-fixed orientation in a fully user-defined manner.
```
...addition of the frame ~~identiers specifief~~ *identifiers specified* above...
(The same error was copy-pasted in the following two paragraphs as well)

gaffarelj commented on docs/source/_src_user_guide/state_propagation/environment_setup/thrust_refactor/thrust_refactor.rst:
```diff
@@ -0,0 +1,168 @@
+
+**propagation_setup.thrust.custom_thrust_direction**  This option is now replaced with the equivalent :func:`~tudatpy.numerical_simulation.environment_setup.rotation_model.custom_inertial_direction_based` option, with the addition of the frame identiers specifief above. This rotation model is very typical for the use of thrust: it imposes that a body-fixed axis (now limited to the body-fixed x-ax ...
+
+**propagation_setup.thrust.thrust_direction_from_state_guidance**  This option is replaced by the :func:`~tudatpy.numerical_simulation.environment_setup.rotation_model.orbital_state_direction_based`. The inputs are equivalent, with the addition of the frame identiers specifief above. Yhis rotation model is a typical example for a basic use of thrust: it imposes that a body-fixed axis (now limited ...
```
~~Yhis~~ *This* rotation model is a....

gaffarelj reviewed (commented):
Great additions and overhaul!
I made a few comments, all about relatively non-critical things.

I like the level of depth used to describe the environment, and I can already see the "custom models" and "use of environment during propagation" sections be massively useful to students doing their thesis (and other research). It eases the current learning curve with clearer explanations and examples.

gaffarelj commented:
(I see that all of my comments actually got sent here as well, sorry for the spam. It's probably clearer to go through them by looking at the [Files changed](https://github.com/tudat-team/tudat-space/pull/71/files))

DominicDirkx commented:
Awesome, thanks!

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/architecture/environment_during_propagation.rst:
```diff
@@ -0,0 +1,131 @@
+
+===================================================
+Interacting with the environment during propagation
+===================================================
```
Top-level comment: I really liked this page overall, and it is extremely useful that you specified how the variables that a user would retrieve are defined by default (for example state w.r.t. global frame origin, etc.). It is something that I myself struggled a lot in the past so very nice to see it clearly documented here :)

DominicDirkx replied:
Thanks! I indeed hope that this page will help with the more advanced use of Tudat

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/architecture/environment_during_propagation.rst:
```diff
@@ -0,0 +1,131 @@
+* Retrieving the current atmospheric density during entry to calculate the required vehicle orientation
+* Retrieving properties that cannot be saved as a dependent variable to calculate a custom dependent variable
+* ...
+
```
here you mention three possible examples for which this documentation page would be useful. The first two are directly addressed by some subsections (how to retrieve translational state / current atmospheric density), while the third is more top-level. I understand why this is here, but it seems to me a bit decorrelated from the page content. Would it make sense to either add a small example showing how this could be done, or add a link to the documentation page where this is explicitely described?

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/architecture/environment_during_propagation.rst:
```diff
@@ -0,0 +1,131 @@
+
+    The current rotational state of a body is defined by its current orientation w.r.t. the global frame orientation (see :ref:`reference_frames_global_orientation`). This orientation is defined by a quaternion (see :ref:`quaternion_definition`), but during a simulation a user will typically interact with the rotation matrix. The rotation matrix from the inertial to body-fixed frame is retrieved  ...
+    
+    The time-derivative of the orientation is provided in two formulations (with equivalent information content): the angular velocity vector of the body-fixed frame, and the time derivative of the rotation matrix. The angular velocity vector, in inertial and body-fixed coordinates, is obtained from the :attr:`~tudatpy.numerical_simulation.environment.Body.inertial_angular_velocity` and :attr:`~t ...
```
> Alternatively, the time-derivative of the rotation matrix from inertial to body-fixed frame is given by :attr:`~tudatpy.numerical_simulation.environment.Body.inertial_to_body_fixed_frame`

Is it the derivative of the rotation matrix that is retrieved here? It seems to me that it is the rotation matrix itself(?)

DominicDirkx replied:
You're right of course :)

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/architecture/environment_during_propagation.rst:
```diff
@@ -0,0 +1,131 @@
+Flight conditions
+-----------------
+
+    The :class:`~tudatpy.numerical_simulation.environment.FlightConditions` class, and its derived class :class:`~tudatpy.numerical_simulation.environment.AtmosphericFlightConditions` stores data relating to altitude, flight angles, local atmospheric properties, etc. The ``FlightConditions`` class is atypical, in the sense that a user does not provide settings for the flight conditions when creat ...
```
small typo: resposible

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst:
```diff
@@ -0,0 +1,129 @@
+
+Specific Considerations
+=======================
+
```
Top level comments:

- I find this section is extremely useful to highlight some important (environment) model choices that are sort of specific to Tudat. As mentioned during last meeting, I think that it would be nice to add a link to this on the main environment page (I know that I already mentioned that, just adding it here as a reminder :) ).
- Overall, I am not entirely convinced by the subsections order. I would suggest to either follow the alphebetical order, or agree on an "order of importance" for the different environment models that we could stick to on **all** environment-related pages. I noticed that the order in which models are discussed is for instance different from the one used in "Interacting with the environment during propagation".
- Concerning the title, I think that we discussed changing it to "Points of attention" during the meeting(?), but I don't know what are your thoughts on this?
- My apologies in advance if the following is only a matter of personal taste/interpretation: If I understand correctly the idea of this page is to highlight some points of attention in order to set up appropriate environment models, and not to discuss all models (dedicated page for this). This is in practice exactly what it does but I have the feeling that it might not always be entirely clear why some models are mentioned here (i.e. what makes them different and therefore worth an additional piece of "warning"). This information is actually there, but may somewhat not be immediately apparent to the new Tudat user. Maybe just reformulating/restructuring the introduction paragraph of each subsection would be enough? What I have in mind is a short sentence highlight which models will be elaborated upon and why. I'm not sure if the above is clear, and maybe this feeling also is not shared by others(?). My apologies if this comment ends up being irrelevant!

DominicDirkx replied:
You don't think "whichever order it popped into my head" is a good paradigm for deciding the order? I'm shocked! Shocked and insulted, I say! ;)

Your last point is a good one. How about:
* Adding a link to the specific subsection of points of attention here: https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst#available-model-types (per relevant model)
* Adding a table of contents with a one line explanation per model here: https://github.com/tudat-team/tudat-space/blob/feature/environment_revision/docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst#specific-considerations

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst:
```diff
@@ -0,0 +1,129 @@
+Rotation models
+---------------
+
+Tudat has a broad range of rotation models available. In principle, these models can be assigned to both celestial bodies and natural bodies. 
```
did you mean artificial and natural bodies?

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst:
```diff
@@ -0,0 +1,129 @@
+However, a subset of these models is typically only applied to natural *or* artificial bodies. Rotation models have a wide range of, sometimes indirect, influences on the dynamics
+
+* A spherical harmonic acceleration exerted by a central body is first evaluated in a body-fixed frame, and the transformed to an inertial frame. Consequently, the central body's rotation has a fundamental influence on the exerted spherical harmonic acceleration
+* A thrust acceleration in Tudat is calculated from two models: (1) an engine model, which defined the body-fixed direction of the thrust, and the magnitude of the thrust (2) the orientation of the body in space, defined by its rotation model
```
small typo: (1) an engine model, which **defines** the body-fixed

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst:
```diff
@@ -0,0 +1,129 @@
+
+Two rotation models, which are typically used for vehicles under thrust, and/or vehicles in an atmosphere, are the following:
+
+* The rotation model :func:`~tudatpy.numerical_simulation.environment_setup.rotation_model.aerodynamic_angle_based`, which calculates the body's rotation based on the angle of attack, sideslip angle and bank angle. Note that these angles are definend w.r.t. the relative wind. This model is typical when using, for instance, a re-entry simulation. It imposes these three angles, and calculates the b ...
```
but _calculates it (?)_ by imposing...

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/create_bodies/available.rst:
```diff
@@ -0,0 +1,129 @@
+* The body for which the state is required *is not* in the ephemeris Spice kernel
+
+In both cases, a user should load additional Spice kernels. This can be done using the :func:`~tudatpy.interface.spice.load_kernel`. Spice kernels for many bodies may be found in a number of places. The 'goto' place for Spice kernels for ephemerides is the NAIF website (developers of Spice), which you can find `here <https://naif.jpl.nasa.gov/pub/naif/generic_kernels/spk/>`_.
+
```
Top-level comment: I am not sure whether this is the best place to mention this (actually I'm quite sure that it is not), but I think that it might be quite useful to add a page or section somewhere stating explicitely what the spice kernels which are loaded by default contain (e.g. ephemerides available from which bodies, etc.). I know that there are also arguments against providing such a list, and that it will requires some updates, but it would address a recurring question from students and it would clarify which body settings are available per default. What are your thoughts about this?

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+
+Propagator pre-processing
+=========================
+
```
this subsection is an extremely useful piece of documentation, and the same goes for the post-processing part!

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+* It is checked whether the propagator settings are feasible (e.g. no body A propagated w.r.t. B, *and* body B propagated w.r.t. body A, etc.).
+* Several objects are created that are used during/after the propagation
+  
+  * An ``EnvironmentUpdater`` object (in C++; not exposed to Python), which decided which environment models need to be updated for each function evaluation (:ref:`single_propagation_evaluation:` and how)
```
bracket misplaced I think

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+  
+  * An ``EnvironmentUpdater`` object (in C++; not exposed to Python), which decided which environment models need to be updated for each function evaluation (:ref:`single_propagation_evaluation:` and how)
+  * A ``ReferenceFrameManager`` object (in C++; not *yet* exposed to Python), which allows translations between ephemerides with different origins to be performed
+  * If the ``set_integrated_result`` to the :class:`~tudatpy.numerical_simulation.SingleArcSimulator` is set to true: A set of ``IntegratedStateProcessor`` objects (in C++; not exposed to Python) which is used to post-process the propagation results, and make any required transformations to reset the ephemerides of the propagated bodies.   
```
perhaps obvious, but maybe worth specifying that it resets the ephemerides of the propagated bodies using their newly propagated state history solutions. "resetting" can be misinterpreted as "back to default"

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+
+.. _single_propagation_evaluation:
+
+A single function evaluation
```
what about changing the title to "What happens during a single function evaluation?" or smtg along this line? I am not entirely happy with this reformulation though, but smtg that I cannot precisely point out triggers me with the current title (same for the "A single time step)

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+* The time-dependent properties of the environment are updated to the current time and propagated state. Only those time/state-dependent models that are needed for either the dynamics or the dependent variables are updated during each time step. For instance, if Jupiter is a body in the environment, but Jupiter's state plays no role in either the dynamics or in the dependent variables that are sa ...
+* Each state derivative model (acceleration, torque, etc.) required for the calculation of the state derivative is evaluated. If variational equations are required, the state derivative partials are evaluated
+* The derivative of each propagated state :math:`\mathbf{x}_{i}` is evaluated from the separate state derivatives (e.g. accelerations are used to compute derivative of Kepler elements, if propagating Kepler elements), and concatenated into the complete state derivative vector :math:`\dot{\mathbf{x}}`
+
```
This might require a bit more work (so maybe for future iterations of the documentation), but it could be nice to illustrate this section (and the single time step one) with some kind of flowchart, for instance using the "propagating keplerian elements of the Moon" example mentioned throughout this paragraph(?)

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+* Check that the independent variable or state are neither infinity nor NaN. If either one is, an exception is thrown.
+* The current state of the environment from the previous time step (translational state; orientation, altitude, etc. of bodies) are cleared, ensuring that they are recomputed for the current time step.  (see :ref:`environment_during_propagation` for the properties of bodies that are set during a time step and how to retrieve them).
+* The propagated state vector :math:`\mathbf{x}` is split into its constituent parts (*e.g.* translational, rotational, *etc.* states of separate bodies that are propagated).
+* The separate propagated states, in 'propagated coordinates', are converted to 'conventional coordinates' (see :ref:`conventional_propagated_states`). For instance: if Kepler elements of the Moon w.r.t.t the Earth are propagated, this step converts those Kepler elements (propagated state) to Cartesian elements (conventional state)
```
small typo _w.r.t._

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+
+Depending on the integrator that is used, a single time step may require one or several function evaluations of the state derivative function :math:`\mathbf{f}`. The full propagation loop, which succesively calls the numerical integrator to advance the state, is in the ``integrateEquationsFromIntegrator`` function (in C++; not exposed to Python). The steps for a single time step are the following ...
+ 
+* Check that the independent variable or state are neither infinity nor NaN. If either one is, the propagation is tagged as being unsuccesful (``nan_or_inf_detected_in_state`` from :class:`~tudatpy.numerical_simulation.propagation.PropagationTerminationReason`) and the results up until the current point are returned.
```
small typo: _unsuccessful_

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+After the propagation is finished, the following post-processing steps are performed before returning the simulation to the user:
+
+* The propagated states are converted to conventional states. After the propagation, the time histories of both may be extracted from the :attr:`~tudatpy.numerical_simulation.SingleArcSimulator.unprocessed_state_history` and :attr:`~tudatpy.numerical_simulation.SingleArcSimulator.state_history` atributes, respectively
+* If the ``set_integrated_result`` to the :class:`~tudatpy.numerical_simulation.SingleArcSimulator` is set to true, the propagated states (in conventional formulation) are used to reset the environment of the propagated body/bodies. For the different state types, this means:
```
Same comment as above for the set_integrated_result functionality: might be worth explicitely stating that the ephemeris is updated to the propagated solution (for new tudat users)?

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/propagation_setup/propagation_architecture.rst:
```diff
@@ -0,0 +1,81 @@
+  * Translational dynamics: the propagated translational state of the body is used to create an interpolator (:func:`~tudatpy.math.interpolators.lagrange_interpolation`, ``number_of_points``=6), which is used to update the :func:`~tudatpy.numerical_simulation.environment_setup.ephemeris.tabulated` ephemeris of the body. If needed, a translation from the propagation origin to the ephemeris origin  ...
+  *  Rotational dynamics: the propagated rotational state of the body is used to create an interpolator (:func:`~tudatpy.math.interpolators.lagrange_interpolation`, ``number_of_points``= 6), which is used to create a tabulated rotation model (not yet exposed to Python). At present, this option is only possible if the propagated body starts out with *no* rotation model. An update to allow the same ...
+  *  Mass dynamics: the propagated mass of the body is used to create an interpolator (:func:`~tudatpy.math.interpolators.lagrange_interpolation`, ``number_of_points``=6), which is used to update the mass function of the body.
+* If the ``clear_numerical_solutions`` to the :class:`~tudatpy.numerical_simulation.SingleArcSimulator` is set to true, the state (processed and unprocessed) and dependent variable history are deleted, *after* having reset the environment (if ``set_integrated_result`` was set to true; see above). In this case, the ephemerides are reset with the propagated dynamics, but the the results of the prop ...
```
I am not sure about TudatPy but it might be worth mentioning (here or elsewhere) that, for Tudat, clear_numerical_solutions is by default set to true for multi-arc propagation. That could also go in the "Point of attention/specific considerations" page for propagation?

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/architecture/custom_settings.rst:
```diff
@@ -0,0 +1,121 @@
+* Custom state :func:`~tudatpy.numerical_simulation.propagation_setup.propagation.custom_state` *Note:* this custom function has current time, and the current custom state, as input
+* Custom termination setting :func:`~tudatpy.numerical_simulation.propagation_setup.propagator.custom_termination`  *Note:* this custom function takes no inputs, if the current time is needed, it can be retrieved from the :attr:`~tudatpy.numerical_simulation.environment.SystemOfBodies.current_time` attribute of the :class:`~tudatpy.numerical_simulation.environment.SystemOfBodies`
+* Custom dependent variable :func:`~tudatpy.numerical_simulation.propagation_setup.dependent_variable.custom_dependent_variable` *Note:* this custom function takes no inputs, if the current time is needed, it can be retrieved from the :attr:`~tudatpy.numerical_simulation.environment.SystemOfBodies.current_time` attribute of the :class:`~tudatpy.numerical_simulation.environment.SystemOfBodies`
+
```
Small question/suggestion: would it be worth adding one example for creating a custom termination setting and/or dependent variable? They are a bit different from the other custom "models" since time is not (necessarily) an input. Actually, one might argue that termination settings/dependent variables are not really "models" and they could be treated independently from the rest of the custom propagation models?

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/architecture/custom_settings.rst:
```diff
@@ -0,0 +1,121 @@
+   :maxdepth: 1
+
+   environment_during_propagation
+
```
Overall I do not have many comments on this page, I really like it and adding the link to the API is really a good idea!

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/architecture/custom_settings.rst:
```diff
@@ -0,0 +1,121 @@
+
+The latter choice permits complex guidance algorithms to be implemented, in which (for instance) the algorithm for the control surface deflection, thrust magnitude and body-fixed thrust direction (thrust vector control) are calculated in a coupled manner.
+
+For most environment models, the custom model can be fully defined by a standalone function , and can be fully defined by a free function (which may of course call other functions from tudat, other packages or your own code. A custom function will, in numerous cases, have to depend on properties of bodies (states, altitude, density, etc.). How to access such properties of the environment *during* ...
```
> (which may of course call other functions from tudat, other packages or your own code

Missing bracket(?)

MarieFayolle commented on docs/source/_src_getting_started/submodules.rst:
```diff
@@ -0,0 +1,43 @@
+==================
+Tudatpy Submodules
+==================
+
```
Don't have much comments on the following, but I think that it is a very good idea to document this :)

MarieFayolle commented on docs/source/_src_getting_started/submodules.rst:
```diff
@@ -0,0 +1,43 @@
+================================================
+
+TODO?
+
```
what kind of examples do you have in mind here?

MarieFayolle commented on docs/source/_src_getting_started/submodules.rst:
```diff
@@ -0,0 +1,43 @@
+* `tudatpy.util <https://py.api.tudat.space/en/latest/util.html>`_ This submodule contains various small pieces of functionality to support the easy post-processing of results generated with Tudatpy. Unlike most of the main Tudatpy submodules (which are written in C++, and exposed to Python), this submodule is written largely in Python
+* `tudatpy.io <https://py.api.tudat.space/en/latest/io.html>`_ This submodule contains various pieces of functionality for file input-output in Tudatpy. Unlike most of the main Tudatpy submodules (which are written in C++, and exposed to Python), this submodule is written largely in Python
+
+The numerical_simulation submodules
```
I guess the idea is to have a dedicated section focusing on each top-level submodules? similar to what you now have for numerical_simulation?

MarieFayolle commented on docs/source/_src_getting_started/submodules.rst:
```diff
@@ -0,0 +1,43 @@
+
+The top-level submodules of tuduatpy are:
+
+* `tudatpy.numerical_simulation <https://py.api.tudat.space/en/latest/astro.html>`_ This submodule contains the interfaces for the primary application of Tudatpy: numerical state propagation and estimation. The functionality in this submodule consists of a large number of interconnected elements that work together as a whole. The usability of separate functions/classes in this submodule *outside* ...
```
small typo: _functionaluty_

MarieFayolle commented on docs/source/_src_user_guide/state_propagation/environment_setup/thrust_refactor/thrust_refactor.rst:
```diff
@@ -0,0 +1,168 @@
+
+The driver behind the lack of backwards compatibility is the following new paradigm:
+
+* **All** methods in which to define a vehicle's orientation (with the exception of the numerical propagation of the vehicle's rotational dynamics) have to happen by *explicitly* setting the vehicle's rotation model. This means that 
```
Maybe this is a nice place to add a few lines about the underlying motivation for such changes? You mention in the introduction that this page should also describe _why_ the thrust functionalities have undergone such refactoring but I'm not sure if this is actually done (maybe I overlooked it though). Anyway, probably no need to add a detailed explanation, just a few sentences should suffice(?)

DominicDirkx replied:
Good point, done :)

MarieFayolle commented:
Hi Dominic,
I finally had the time to go through your documentation changes. I made a few comments above (not sure that they are all relevant, feel free to disregard those that you don't like). Overall I really like it, congrats for the hard work! In particular, the propagation setup and specific considerations on environment models are great additions I think

DominicDirkx commented on docs/source/_src_getting_started/submodules.rst:
```diff
@@ -0,0 +1,43 @@
+* `tudatpy.numerical_simulation <https://py.api.tudat.space/en/latest/astro.html>`_ This submodule contains the interfaces for the primary application of Tudatpy: numerical state propagation and estimation. The functionality in this submodule consists of a large number of interconnected elements that work together as a whole. The usability of separate functions/classes in this submodule *outside* ...
+* `tudatpy.astro <https://py.api.tudat.space/en/latest/astro.html>`_ This submodule contains various (semi-)standalone functions for astrodynamics applications, which can be used very well outside of a Tudat application/propagation. Submodules contain lists of frame conversions, element conversion, elementary orbit calculations, *etc.*.
+* `tudatpy.trajectory_design <https://py.api.tudat.space/en/latest/trajectory_design.html>`_ This submodule contains functionality for the preliminary design of a full (transfer) orbit, using for instance a Muliple Gravity Assist (MGA) or a low-thrust system. It relies of functionality in the ``astro`` submodule. It is largely independent of the ``numerical_simulation`` submodule, but does contai ...
+* `tudatpy.math <https://py.api.tudat.space/en/latest/math.html>`_  This submodule contains various functions and classes for purely mathematical operations, such as numerical interpolation, numerical integration, *etc.*.
```
That's a good point, I think we should do this at the same point as the upcoming overhaul of the propagator settings, where one of the things I want to do is put integrator settings inside propagator settings. I think this approach will be doable in a backwards-compatible manner, with deprecation warnings.

DominicDirkx commented on docs/source/_src_getting_started/submodules.rst:
```diff
@@ -0,0 +1,43 @@
+* `tudatpy.astro <https://py.api.tudat.space/en/latest/astro.html>`_ This submodule contains various (semi-)standalone functions for astrodynamics applications, which can be used very well outside of a Tudat application/propagation. Submodules contain lists of frame conversions, element conversion, elementary orbit calculations, *etc.*.
+* `tudatpy.trajectory_design <https://py.api.tudat.space/en/latest/trajectory_design.html>`_ This submodule contains functionality for the preliminary design of a full (transfer) orbit, using for instance a Muliple Gravity Assist (MGA) or a low-thrust system. It relies of functionality in the ``astro`` submodule. It is largely independent of the ``numerical_simulation`` submodule, but does contai ...
+* `tudatpy.math <https://py.api.tudat.space/en/latest/math.html>`_  This submodule contains various functions and classes for purely mathematical operations, such as numerical interpolation, numerical integration, *etc.*.
+* `tudatpy.interface <https://py.api.tudat.space/en/latest/interface.html>`_ This submodule contains functionality to interface with various external packages which Tudat uses, such as `Spice <https://naif.jpl.nasa.gov/naif/toolkit.html>`_
```
Good point, I had not considered this, but it sounds like a good idea. Could you open an issue on this to remind us to do this in thee future? We could also discuss this during our next development meeting.

gaffarelj replied:
Perfect, I submitted this as [tudatpy issue 75](https://github.com/tudat-team/tudatpy/issues/75).
